### PR TITLE
Hsy docs 11372

### PR DESCRIPTION
**Pull request by shinyou21**
code for https://jira.mongodb.org/browse/DOCS-11372
: ArrayQuery.java contains methods that illustrates the use of the $in operator; use of the $all operator; an equality match on an array field, use of the $elemMatch operator.
: Two of the methods make use of projection.
: ArrayQueryTest.java in testsrc contains the unit tests for all methods + test for connection to the cluster.

### CODE REVIEW

File: ArrayQuery/src/ArrayQuery.java [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	}
+	
+	
+	/**
+	 * Retrieves documents filtered by shipping address city. use of $in
+	 * @param connectionString: To MongoDB instance/MongoDB Cluster.
+	 * @param cities: array of Shipping address cities to filter documents.
+	 * @return List<Document>: ArrayList of matching documents.
+	 */
+	public static List<Document> InOperator(String connectionString, String[] cities)
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Sample apps are more interesting if there is a narrative around them. In general this app should look like a portion of an e-commerce app with method names that make sense in that context.

In general, method names should describe what the method does. This method filters shipping addresses by city. The name should reflect that.

File: ArrayQuery/src/ArrayQuery.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+		MongoClientURI clientUri = new MongoClientURI(connectionString);
+		try(MongoClient client = new MongoClient(clientUri))
+		{
+		
+			MongoDatabase database = client.getDatabase("stores");
+			MongoCollection<Document> collection = database.getCollection("orders");		
+			
+			//List<Document> queryResult = collection.find(in("shippingAddress.city","Floris", "Calpine", "Orason", "Allentown")).projection(fields(include("subtotal","shipping","shippingAddress.city"),excludeId())).into(new ArrayList<Document>());
+			List<Document> queryResult = collection.find(in("shippingAddress.city",cities)).projection(fields(include("subtotal","shipping","shippingAddress.city"))).into(new ArrayList<Document>());
+
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Please format the above so that it is more readable. You'll need to break this apart onto several lines. Remember that code in these apps needs to be easy to understand.

File: ArrayQuery/src/ArrayQuery.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+			return null;
+		}		
+	}
+	/**
+	 * Retrieves documents filtered by shipping items. use of $all
+	 * @param connectionString: To MongoDB instance/MongoDB Cluster.
+	 * @param items: array of items to filter documents.
+	 * @return List<Document>: ArrayList of matching documents.
+	 */
+	public List<Document> AllOperator(String connectionString,String[] items)
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Needs a more descriptive name.

File: ArrayQuery/src/ArrayQuery.java [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	{
+		
+		MongoClientURI clientUri = new MongoClientURI(connectionString);
+		try(MongoClient client = new MongoClient(clientUri))
+		{
+			MongoDatabase database = client.getDatabase("stores");
+			MongoCollection<Document> collection = database.getCollection("orders");		
+		
+			List<Document> queryResult = collection.find(all("lineitems.name",items)).projection(fields(include("subtotal","lineitems.name","shippingAddress.city"))).into(new ArrayList<Document>());
+			return queryResult;
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Please make more readable.

File: ArrayQuery/src/ArrayQuery.java [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+		}
+	}
+	
+	/**
+	 * Retrieves documents filtered by status to retrieve data from arrays
+	 * @param connectionString: To MongoDB instance/MongoDB Cluster.
+	 * @param status: string value of status to filter documents.
+	 * @return List<Document>: ArrayList of matching documents.
+	 */
+	public static List<Document> ArrayEqualityOperator(String connectionString, String status)
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Needs a more descriptive name.

File: ArrayQuery/src/ArrayQuery.java [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+	
+	
+	/**
+	 * Retrieves documents filtered by items and their quantity. use of $elemMatch
+	 * @param connectionString: To MongoDB instance/MongoDB Cluster.
+	 * @param item_name: item name to filter documents.
+	 * @param quantity: integer value to filter documents that greater than this value.
+	 * @return List<Document>: ArrayList of matching documents.
+	 */
+	public List<Document> ElemMatchOperator(String connectionString, String item_name, int quantity)
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: More descriptive name.

File: ArrayQuery/src/ArrayQuery.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+	{
+		
+		MongoClientURI clientUri = new MongoClientURI(connectionString);
+		try(MongoClient client = new MongoClient(clientUri))
+		{
+			MongoDatabase database = client.getDatabase("stores");
+			MongoCollection<Document> collection = database.getCollection("orders");		
+			
+			List<Document> queryResult = collection.find(elemMatch("lineitems",and(eq("name",item_name),gt("quantity",quantity)))).into(new ArrayList<Document>());
+
```
Comment (shannonbradshaw) [inline ArrayQuery/src/ArrayQuery.java]: Please make the above line of code more readable by breaking into multiple lines.

File: ArrayQuery/testsrc/ArrayQueryTest.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+	@Test
+	public void TestConnection() {		
+		MongoClientURI clientUri = new MongoClientURI(CONNECTION);
+		try(MongoClient client = new MongoClient(clientUri))
+		{
+			assertNotNull(client);
+		}    
+	}
+	@Test
+	public void InOperatorTest() {
```
Comment (shannonbradshaw) [inline ArrayQuery/testsrc/ArrayQueryTest.java]: This test could be considerably more robust and really should be. You will want to make sure that it works (and does not throw an exception) for corner cases such as no matching results or an empty array passed as a part of the filter. Are there other corner cases we should consider?

Also, it would be ideal to test that the documents returned are the ones you expect rather than just count the results. This ensures that if our writeup of examples in the docs includes example output, that the code will provide the correct output.

**Comment by shannonbradshaw** [comment]
@shinyou21 I finished my code review, please have a look at my comments. Let me know when you're ready for me to do another pass.

**Comment by shinyou21** [comment]
@shannonbradshaw, I have made changes to the test cases

**Review by shinyou21 — COMMENTED** [review]
@shannonbradshaw just checking in to see if the changes I have made are alright

File: ArrayQuery/src/ArrayQuery.java (hunk 2 above)
Comment (shannonbradshaw) [reply ArrayQuery/src/ArrayQuery.java]: Looks good to me (LGTM)

**Review by shannonbradshaw — APPROVED** [review]
Your changes look good to me (LGTM)!

**Comment by shannonbradshaw** [comment]
Last changes LGTM
